keps/sig-cli/kubectl-socket-support.md

```diff
@@ -0,0 +1,195 @@
+  - "@answer1991"
+owning-sig: sig-cli
+participating-sigs:
+reviewers:
+  - "@pwittrock"
+  - "@soltysh"
+  - "@seans3"
+approvers:
+  - "@pwittrock"
+  - "@soltysh"
+  - "@liggitt"
+editor: "@dixudx"
+creation-date: 2020-03-16
+last-updated: 2020-03-16
+status: implementable
+---
+
+# Socket Support for Kubectl
+
+## Table of Contents
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
+    - [Beta -&gt; GA Graduation](#beta---ga-graduation)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location
+in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+This proposal adds a new feature that would allow sockets (`unix`, `tcp` and `fd`) as a new protocol for the server to be connected.
+
+## Motivation
+
+Currently several kinds of credentials (passwords, tokens, x509 certificates and etc) are stored in a kubeconfig file and
+persisted on disk. It is not safe and controllable, especially when this kubeconfig file with credentials gets spread to everywhere.
+Moreover, it will become difficult to revoke the access to the apiserver.
+
+Sockets are used as a secured proxy into some infrastructure. For now kubectl does not expose a way to send the request
+over a unix socket. Currently we've only got `http` and `https`. `Docker` client does provide such supports via three
+different types of Socket: `unix`, `tcp`, and `fd`.
+
+It would be great to be able to send requests from kubectl over sockets.
```

Comment:
does the kube-apiserver have the capacity to serve on a socket? does this proposal make sense without that?

Reply:
No, kube-apiserver can't do this and I don't see us adding it.

However if another copy of kubectl is proxying onto a unix domain socket...

Reply:
> does the kube-apiserver have the capacity to serve on a socket?

Not have to for now.

> does this proposal make sense without that?

Yes. kubectl can use unix socket to transport over HTTP.